docs/lsp.md

```diff
@@ -0,0 +1,489 @@
+
+participant Client
+participant FsLSP
+participant WorkspaceStateManagement
+participant FCS
+
+Client->>+FsLSP: textDocument/diagnostic
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+FsLSP->>FCS: ParseAndCheckFileInProject
+activate FCS
+Client->>FsLSP: textDocument/didChange
+FsLSP->>FCS: Cancel ParseAndCheckFileInProject
+deactivate FCS
+FsLSP->>WorkspaceStateManagement: Update workspace state
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+
+FsLSP->>+FCS: ParseAndCheckFileInProject
+FCS-->>-FsLSP: FSharpCheckFileResults
+FsLSP-->>-Client: DocumentDiagnosticReport
+
+```
+
+</details>
+
+
+#### Request tracking
+
+This will hopefully be handled by the LSP library. Although we might need to be able to keep track of requests in progress for the purpose of cancellation.
+
+#### Extensibility
+
+The FsLSP library should be extensible by adding custom endpoints or overriding existing ones.
+
+It should also allow to hook into background processes or events to customize behavior.
+
+### NewVsix
+
+New VS extension that will be backed by the LSP server. It should mostly contain plumbing to connect LSP to the editor and CPS.
+
+> Is there a library available to simplify some of this?
+
+> Will we be able to build this on top of the new VisualStudio.Extensibility?
+
+It should be able to run side by side with the current extension gradually take over the functionality we will add to LSP.
+
+Depending on the extensibility model we will either use the FsLSPServer executable or just load the library directly.
+
+### FCS
+
+The only strict requirement on FCS is to be able to process open files from memory and non-open files from filesystem. But probably best to abstract that within FsLSP and just provide `ISourceText` objects for both cases.
+
+At the moment this is possible via the experimental `DocumentSource.Custom` parameter. But hopefully we will be able to use a new `FSharpProjectSnapshot` model (from [#15179](https://github.com/dotnet/fsharp/pull/15179)) which combines all the necessary inputs into a single object.
+
+
+## LSP library
+
+The library needs to support all the features we want to implement including the ability to add custom F# ones. It also shouldn't incur an unnecessary performance overhead.
+
+### Candidates
+
+| Library  | Pros | Cons |
+| ---  | --- | --- |
+| We write our own | + Complete control over everything | - A lot of extra work |
+| [Ionide LanguageServerProtocol](https://github.com/ionide/LanguageServerProtocol) | + Battle tested with Ionide and FSAC + Natural F# API | - Might not be able to fully customize it to our need |
+| [Roslyn CLaSP](https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/README.md) | + Should be performant + Future support | - Not stable API - Unnatural C# API |
```

Comment:
I guess based on discussion with Rolsyn folks, we'll likely go with CLaSP as base + maybe Ionide for types and additional helpers? We also should, for now, just ask EngSys to upload Ionide package to `dotnet-public`, so we can iterate faster.

Reply:
It probably makes sense to try to prototype with it, see how awkward the API is from F#. Roslyn should also have the types somewhere which we could use, might be good enough for what we need them for.

Reply:
I mean the whole API is just registering handlers and types for messages, and that's it. It feels pretty slim and natural in F#. Roslyn types are not public I believe.